Implement validation and stream utilities

koder95 opened the pull request:
Added CandidateValidator with custom validation logic. Implemented various StreamPractice methods for processing lists, including finding minimum even numbers, calculating averages, filtering people by age and sex, extracting cat names, and validating candidates.

koder95 committed:
Implement validation and stream utilities

Added CandidateValidator with custom validation logic. Implemented various StreamPractice methods for processing lists, including finding minimum even numbers, calculating averages, filtering people by age and sex, extracting cat names, and validating candidates.

mateacademy-ai-mentor commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,30 @@
+
+    @Override
+    public boolean test(Candidate candidate) {
+        boolean isUkrainian = candidate.getNationality().equals("Ukrainian");
```
This violates checklist item #6: "Use constants \nMagic numbers and strings decrease code readability, let's avoid it and use constants in class CandidateValidator". The code uses the string literal `"Ukrainian"` directly in the test method instead of declaring a `private static final String` constant and using it (see line below). Please extract the nationality into a constant (for example `private static final String REQUIRED_NATIONALITY = "Ukrainian";`) and use that constant in the equality check to satisfy the checklist requirement.

mateacademy-ai-mentor commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -14,7 +16,12 @@ public class StreamPractice {
      */
     public int findMinEvenNumber(List<String> numbers) {
-        return 0;
+        return numbers.stream()
+                .flatMap(numbs -> Arrays.stream(numbs.split(",")))
+                .mapToInt(Integer::parseInt)
```
The token parsing pipeline does not trim whitespace before parsing. The task description includes inputs like `{"5,30,100", "0, 22, 7", ...}` which contain spaces after commas; the code currently does `Arrays.stream(numbs.split(","))` then `Integer::parseInt` which will throw or parse incorrectly for tokens with surrounding spaces. Ensure you trim each token before parsing (for example map tokens with `String::trim` before `Integer::parseInt`). See the task description example input for whitespace after commas.

mateacademy-ai-mentor commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -14,7 +16,12 @@ public class StreamPractice {
+                .mapToInt(Integer::parseInt)
+                .filter(i -> i % 2 == 0)
+                .min()
+                .orElseThrow(() -> new RuntimeException("Can't get min value from list"));
```
The thrown RuntimeException message must match the exact required text. The checklist states: In method `findMinEvenNumber()` exception message should look like this:
"Can't get min value from list: " + numbers
Current code throws `new RuntimeException("Can't get min value from list")` (missing the input list). Update to include the `numbers` list exactly as required, e.g. `orElseThrow(() -> new RuntimeException("Can't get min value from list: " + numbers))`. Quote from checklist: In method `findMinEvenNumber()` exception message should look like this:
"Can't get min value from list: " + numbers.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Requesting changes due to critical blockers that violate task requirements and can break functionality.

Critical issues to fix:
1. StreamPractice.findMinEvenNumber (around the parsing pipeline): You split by commas but do not trim each token before parsing. Inputs like "0, 22, 7" will fail or parse incorrectly. Trim each token before Integer.parseInt (e.g., map with String::trim). [CHECKLIST ITEM]
2. StreamPractice.findMinEvenNumber exception message: The thrown RuntimeException must be exactly: "Can't get min value from list: " + numbers. Currently it’s missing the list. Update your orElseThrow to include the input list exactly as specified. This is required and will cause tests to fail if not exact. [CHECKLIST ITEM]
3. CandidateValidator.java: You use the magic string "Ukrainian" inline (see the nationality equality check). Per the checklist, extract this into a constant, e.g., private static final String REQUIRED_NATIONALITY = "Ukrainian" and use that constant in the comparison. [CHECKLIST ITEM #6]

Minor suggestions (nice-to-have): Use REQUIRED_NATIONALITY.equals(candidate.getNationality()) to avoid NPEs and consider trimming/validating period tokens before Integer.parseInt for robustness.

Overall, your stream usage and core logic look good—once the above are addressed, this should be ready to pass. Keep going—you’re close!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

koder95 committed:
Refactor candidate validation and stream parsing

Extracted nationality string to a constant in CandidateValidator for better maintainability. Improved StreamPractice.findMinEvenNumber by trimming input strings and enhancing exception message with input context.

koder95 committed:
Refactor imports and improve code readability

Organized import statements in CandidateValidator and StreamPractice for clarity. Introduced a constant for exception messages and improved formatting in StreamPractice to enhance readability and maintainability.

mateacademy-ai-mentor commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -35,7 +49,10 @@ public Double getOddNumsAverage(List<Integer> numbers) {
     public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getSex() == Person.Sex.MAN)
+                .filter(person -> person.getAge() >= fromAge && person.getAge() <= toAge)
```
This violates checklist item: "Try to use a minimum amount of stream operations" .

Explanation: `selectMenByAge` uses two consecutive .filter(...) operations (one checking `sex`, another checking `age`). Combine these into a single filter (or use a named Predicate) to reduce stream operations and follow the checklist guidance. Example approach: `.filter(person -> person.getSex() == Person.Sex.MAN && person.getAge() >= fromAge && person.getAge() <= toAge)` (or extract that predicate). See the method in StreamPractice for context .

mateacademy-ai-mentor commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -50,7 +67,11 @@ public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toA
                                           int maleToAge, List<Person> peopleList) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getAge() >= fromAge)
+                .filter(person -> person.getSex() == Person.Sex.MAN && person.getAge() <= maleToAge
+                        || person.getSex() == Person.Sex.WOMAN && person.getAge() <= femaleToAge)
```
This violates checklist item: "Try to use a minimum amount of stream operations" .

Explanation: `getWorkablePeople` uses two .filter(...) calls: one for `person.getAge() >= fromAge` and a second with the combined sex/age logic. These can be merged into a single filter (or a named Predicate) to minimize stream operations and satisfy the checklist. For example, combine the conditions so the stream uses a single .filter(...) rather than two separate filters. Refer to the method in StreamPractice for the exact location .

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

**Decision: REQUEST_CHANGES**

Your solution is very close and most requirements are met, but there is a checklist violation that must be addressed before approval: minimizing stream operations.

Critical issues to fix (checklist blockers):
- In `src/main/java/practice/StreamPractice.java`, method `selectMenByAge`: you use two consecutive `filter(...)` calls (sex and age). This violates the checklist item “Try to use a minimum amount of stream operations.” Combine them into a single predicate, e.g., `.filter(p -> p.getSex() == Person.Sex.MAN && p.getAge() >= fromAge && p.getAge() <= toAge)`. This reduces operations and keeps the logic clear.
- In the same file, method `getWorkablePeople`: there are two `filter(...)` calls (`person.getAge() >= fromAge` and then the sex/upper-bound check). Merge them into one filter so the stream has a single filtering step. A named `Predicate<Person>` is also acceptable to keep readability while meeting the checklist.

Minor suggestion (non-blocker): Your `CandidateValidator` works as required, but consider trimming inputs and handling multiple date ranges in `periodsInUkr` for robustness.

Once you merge the filters as described, the PR should meet all requirements. Good work overall—just a small adjustment to align with the checklist!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

koder95 committed:
Refactor stream filters for improved readability

Combined multiple filter operations into single expressions in selectMenByAge and getWorkablePeople methods for better readability and efficiency.

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Your solution is well-structured and meets the task requirements using the Stream API effectively. You correctly implemented all methods, used combined predicates, compared enums with `==`, applied boolean operators `&&`/`||`, and encapsulated candidate checks in `CandidateValidator` with meaningful constants. The exception messages and sorting behavior are correct, and your stream formatting is clean and readable.

Educational override note: One optional robustness improvement — in `CandidateValidator` when parsing `periodsInUkr` (e.g., around the split on `-`), consider trimming parts and safely handling unexpected formats or multiple ranges to avoid `NumberFormatException` or edge-case failures. For example, normalize input like `"1999 - 2015"` with `trim()` before `Integer.parseInt`.

Overall, great job applying the checklist and producing concise, readable code. Keep up the solid work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

MateuszRostek commented on src/main/java/practice/CandidateValidator.java:
```diff
@@ -1,5 +1,30 @@
+    private static final int LIVING_IN_UKR_REQUIREMENT = 10;
+    private static final String REQUIRED_NATIONALITY = "Ukrainian";
+
+    @Override
+    public boolean test(Candidate candidate) {
+        boolean isUkrainian = REQUIRED_NATIONALITY.equals(candidate.getNationality());
+        boolean isOlderThan35 = candidate.getAge() >= AGE_REQUIREMENT;
+        boolean allowedToVote = candidate.isAllowedToVote();
+        String[] split = candidate.getPeriodsInUkr().split("-");
+        int startYear = Integer.parseInt(split[0]);
+        int endYear = Integer.parseInt(split[1]);
+        int inUkr = endYear - startYear;
+        boolean living10YearsInUkr = inUkr >= LIVING_IN_UKR_REQUIREMENT;
+        return isOlderThan35 && allowedToVote && isUkrainian && living10YearsInUkr;
+    }
```
Too much variables, can be simplified to something like this:

```suggestion
    @Override
    public boolean test(Candidate candidate) {
        String[] yearsRange = candidate.getPeriodsInUkr().split("-");
        int stayDuration = Integer.parseInt(yearsRange[1]) - Integer.parseInt(yearsRange[0]);

        return candidate.isAllowedToVote() 
                && candidate.getAge() >= MINIMUM_AGE
                && candidate.getNationality().equals(REQUIRED_NATIONALITY)
                && stayDuration >= MINIMUM_RESIDENCY_YEARS;
    }
```

MateuszRostek commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -23,7 +34,10 @@ public int findMinEvenNumber(List<String> numbers) {
+        return IntStream.range(0, numbers.size())
+                .map(i -> i % 2 == 1 ? numbers.get(i) - 1 : numbers.get(i))
+                .filter(i -> i % 2 == 1)
+                .average().orElseThrow(NoSuchElementException::new);
```
```suggestion
                .average()
                .orElseThrow(NoSuchElementException::new);
```

MateuszRostek commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -35,7 +49,10 @@ public Double getOddNumsAverage(List<Integer> numbers) {
      */
     public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toAge) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getSex() == Person.Sex.MAN
+                        && person.getAge() >= fromAge && person.getAge() <= toAge)
+                .toList();
```
```suggestion
        return peopleList.stream()
                .filter(p -> p.getAge() >= fromAge
                        && p.getAge() <= toAge
                        && p.getSex() == Person.Sex.MAN)
                .toList();
```

MateuszRostek commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -50,7 +67,11 @@ public List<Person> selectMenByAge(List<Person> peopleList, int fromAge, int toA
     public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
                                           int maleToAge, List<Person> peopleList) {
-        return Collections.emptyList();
+        return peopleList.stream()
+                .filter(person -> person.getAge() >= fromAge
+                        && (person.getSex() == Person.Sex.MAN && person.getAge() <= maleToAge
+                        || person.getSex() == Person.Sex.WOMAN && person.getAge() <= femaleToAge))
+                .toList();
```
```suggestion
        return peopleList.stream()
                .filter(p -> p.getAge() >= fromAge
                        && (p.getSex() == Person.Sex.MAN
                        ? p.getAge() <= maleToAge
                        : p.getAge() <= femaleToAge))
                .toList();
```

MateuszRostek commented on src/main/java/practice/StreamPractice.java:
```diff
@@ -59,7 +80,12 @@ public List<Person> getWorkablePeople(int fromAge, int femaleToAge,
+        return peopleList.stream()
+                .filter(person -> person.getAge() >= femaleAge
+                        && person.getSex() == Person.Sex.WOMAN)
+                .flatMap(person -> person.getCats().stream())
```
```suggestion
                .flatMap(Collection::stream)
```

MateuszRostek reviewed (approved):
Good job! See minor advice

koder95 committed:
Update src/main/java/practice/StreamPractice.java

Co-authored-by: Mateusz Rostek <140905715+MateuszRostek@users.noreply.github.com>
